Conversation saving into jupyter notebook

MauroAbidalCarrer opened the pull request:
# Conversation saving into jupyter notebook
Hello, this is a pull request to  allow the user to save his conversations into jupyter notebooks.  

### Usage:
Add the `--notebook` or `-n` option when running `python3 web_ui.py`:
```shell 
python3 web_ui.py -n path_to_notebook
```
For every new message the notebook will be updated.

### Implementation
I have added a serialization module that uses nbformat package.
The module offers a few functions to serialize code, code outputs/errors, images and usr/assistant messages into the notebook.
The module also uses argparse to allow the user to specify the path to the notebook.
The conversation doesn't get serialized if the notebook option is not used,  
leaving the behavior of `Local-Code-Interpreter`  unchanged if the notebook option is not used.

I am  currently working on notebook deserialization into conversation to allow you pick up where left off.
I wanted to make this PR as soon as the serialization is done so the PR doesn't get to big or out of sync.

I hope this will be useful!

[serialization_demo_gif](https://github.com/MrGreyfun/Local-Code-Interpreter/assets/60320694/e556a375-cec7-42df-ab2b-0414f5e77a40)

MrGreyfun commented on src/response_parser.py:
```diff
@@ -1,6 +1,6 @@
 from abc import ABCMeta, abstractmethod
 from functional import *
-
+# from notebook_serializer import serialize_conv_into_notebook
```
Remove such redundant comments in the code.

MrGreyfun commented:
Hello, thank you for this PR and your efforts to improve this project. One of my development plans for this project is to save the conversion history to Jupyter notebook files, so I'm delighted to see this implementation. However, there are several issues within the code that require further modification before merging:  
   
1. There are some problems with the newly added code. I've added comments to these, please consider making the necessary modifications.  
2. The generated Jupyter notebook file is unable to display non-ascii characters. To rectify this, please use UTF-8 encoding when saving the file.  
3. Please add information on how to save conversions to Jupyter notebooks in the English version of the README.md file.  
   
Additionally, consider adding this GIF image ([serialization_demo_gif](https://github.com/MrGreyfun/Local-Code-Interpreter/assets/60320694/e556a375-cec7-42df-ab2b-0414f5e77a40)) to the README file for a quick preview.

Once these adjustments are made, the PR can be considered for merging.

MrGreyfun commented:
Deserialization need to restore the Python runtime, which I think is the hardest part.

MauroAbidalCarrer commented:
Sure, I will take care of this in a few hours.
:+1:

MauroAbidalCarrer commented:
> which I think is the hardest

I assume that we could simply rerun all the code cells?
Alternatively, there might be some way of serialize the Python runtime as well?

I have already started implementing deserialization of the conversation history.
Once this pull request gets merged we could discuss on the desirialization part further, maybe in an issue.

MauroAbidalCarrer commented on src/bot_backend.py:
```diff
@@ -183,14 +189,25 @@ def add_file_message(self, path, bot_msg):
         )
 
     def add_function_call_response_message(self, function_response: str, save_tokens=True):
+        # Add code to notebook cells.
```
When trying to import either `get_code_str` or `parse_json`, I run into `ImportError` due to circular import.
This is caused by  `from bot_backend.py import *` in `functional.py` and `from functional import *` in response_parser.py` .
Since `response_parser` imports `functional` and `functional` imports `bot_backend` importing either function/method results in a circular import.

I would suggest moving `parse_json` into `response_parser` since it is the only place where it is used and replace from import.

As a quick fix I am importing `parse_json` in the method `add_function_call_response_message` itself.

MrGreyfun replied:
You can use `from functional import parse_json` instead of importing the entire module.

MrGreyfun replied:
Sorry about that, this cannot solve the `ImportError` issue. Please use your quick fix first. I will reorganize my code as soon as I have time.

MauroAbidalCarrer replied:
I used `parse_json` but when `self.function_args_str` is not in json format, calling `parse_json` raises an exception:
```
Traceback (most recent call last):
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/routes.py", line 442, in run_predict
    output = await app.get_blocks().process_api(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/blocks.py", line 1392, in process_api
    result = await self.call_function(
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/blocks.py", line 1111, in call_function
    prediction = await utils.async_iteration(iterator)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 346, in async_iteration
    return await iterator.__anext__()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 339, in __anext__
    return await anyio.to_thread.run_sync(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 322, in run_sync_iterator_async
    return next(iterator)
           ^^^^^^^^^^^^^^
  File "/home/mauro/.local/lib/python3.11/site-packages/gradio/utils.py", line 691, in gen_wrapper
    yield from f(*args, **kwargs)
  File "/home/mauro/projects/local_code_interpreter_fork/src/web_ui.py", line 112, in bot
    exit(-1)
  File "<frozen _sitebuiltins>", line 26, in __call__
SystemExit: -1
```
Which is weird because I have a try/except to catch any error:
```python
        from functional import parse_json
        try:
            code_str = parse_json(self.function_args_str, True)
            if code_str:
                add_code_cell_to_notebook(code_str)
        except Exception as e:
            print(f"Caught exception while trying to add code to notebook: {e}")
```

On the other hand I never get any issue when using `json.loads`.
I see in the docstring of `parse_json` that the reason for the implementation of this warper around `json.loads` are the 
`\\n`s in the string but this issue as been solved with the latest `1106` versions of GPT chat completion.

I faced a similar issue when using the older versions in previous projects.
To prevent this potential case where the string is in *erronous* json, I suggest we simply apply a `replace('\\n', '\n')`
on `function_args_str` before callin `json.loads` and leave the rest as is.

What do you say?

MrGreyfun replied:
In the following code, `s` is an erroneous JSON sting that I encountered in a previous test. However, applying `replace('\\n', '\n')` to it alone doesn't allow successful parsing using `json.loads`.

```python
import json

s = '''{
    "code": "
import numpy as np

x = np.random.rand(100, 100)
y = np.random.rand(100, 100)

z = np.dot(x, y)
z
"
}'''

result = json.loads(s.replace('\\n', '\n'))
```

To minimize code duplication, I've added a `code_str` attribute to the `BotBackend` class. You can rebase your branch and use `self.code_str` to obtain the code string.

MauroAbidalCarrer commented:
@MrGreyfun 
> Hello, thank you for this PR and your efforts to improve this project. One of my development plans for this project is to save the conversion history to Jupyter notebook files, so I'm delighted to see this implementation. However, there are several issues within the code that require further modification before merging:
> 
> 1. There are some problems with the newly added code. I've added comments to these, please consider making the necessary modifications.
> 2. The generated Jupyter notebook file is unable to display non-ascii characters. To rectify this, please use UTF-8 encoding when saving the file.
> 3. Please add information on how to save conversions to Jupyter notebooks in the English version of the README.md file.
> 
> Additionally, I think this GIF image ([serialization_demo_gif](https://github.com/MrGreyfun/Local-Code-Interpreter/assets/60320694/e556a375-cec7-42df-ab2b-0414f5e77a40)) can be added to the README file as a quick preview.
> 
> Once these adjustments are made, the PR can be considered for merging.

I made a few commits:
- Serialization now handles `UTF-8`
- Updated the `README.md`
- Added `replace('\\n', '\n')` to `function_args_str` when calling `json.loads` to protect it from incorrect json generation.
 Let me know if this fix is good for you.

MauroAbidalCarrer commented:
I have rebased the branch and by using self.code_str, it works like a charm :+1: 
Are we ready to merge?

MrGreyfun commented:
> I have rebased the branch and by using self.code_str, it works like a charm 👍 Are we ready to merge?

Yes, I've merged this PR. Thanks again for your effort.
